#01. Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

#02. Importing Data

In [2]:
# Turn project folder path into a string
path = r'/Users/thomasradio/Documents/Jupyter/230621 Tom Radio Achievement 4 Project'

In [3]:
# Check path
path

'/Users/thomasradio/Documents/Jupyter/230621 Tom Radio Achievement 4 Project'

## 3. Import your analysis libraries, as well as your new customer data set as a dataframe.

In [4]:
# Import orders_products_merged_4_8_Task dataset
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_4_8_Task.pkl'))

In [5]:
ords_prods_merge.head()

order_id  user_id  user_order_number  orders_day_of_week  \
0   2539329        1                  1                   2   
1   2398795        1                  2                   3   
2    473747        1                  3                   3   
3   2254736        1                  4                   4   
4    431534        1                  5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered _merge  ... prices    price_range_loc        busiest_days  \
0          0   both  ...    9.0  Mid-range product      Regularly busy   
1          1   both  ...    9.0  Mid-range product  Least busiest days   
2          1   both  ...    9.0  Mid-range product  Least busiest days   
3          1   both  ...    9.0  Mid-range product  Least busiest days   
4          1   both  ...    9.0  Mid-range product  Least busiest days   

   busiest_period_of_day max_order  loyalty_flag avg_price_per_order  \
0         Average orders        10  New customer            6.367797   
1         Average orders        10  New customer            6.367797   
2            Most orders        10  New customer            6.367797   
3         Average orders        10  New customer            6.367797   
4            Most orders        10  New customer            6.367797   

   spending_flag median_days_since_prior_order   order_frequency_flag  
0    Low spender                          20.5  Non-frequent customer  
1    Low spender                          20.5  Non-frequent customer  
2    Low spender                          20.5  Non-frequent customer  
3    Low spender                          20.5  Non-frequent customer  
4    Low spender                          20.5  Non-frequent customer  

[5 rows x 23 columns]

In [6]:
ords_prods_merge.shape

(32404859, 23)

In [7]:
# Import customers dataset
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [8]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
df_cust.shape

(206209, 10)

## 4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [10]:
# Rename columns
df_cust.rename(columns = {'n_dependants' : 'number_of_dependants'}, inplace = True)
df_cust.rename(columns = {'fam_status' : 'family_status'}, inplace = True)

In [11]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependants family_status  income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

## 5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

In [12]:
# Look for missing values in each column
df_cust.isnull().sum()

user_id                     0
First Name              11259
Surnam                      0
Gender                      0
STATE                       0
Age                         0
date_joined                 0
number_of_dependants        0
family_status               0
income                      0
dtype: int64

In [13]:
# Look for missing values in 'First Name' column
df_cust['First Name'].isnull().value_counts(dropna = False)

False    194950
True      11259
Name: First Name, dtype: int64

### There are 11,259 missing values in the First Name column. I left these alone since I can't make up a first name, and the rest of the information in these rows will still be valuable.

In [14]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   First Name            194950 non-null  object
 2   Surnam                206209 non-null  object
 3   Gender                206209 non-null  object
 4   STATE                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   number_of_dependants  206209 non-null  int64 
 8   family_status         206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [15]:
df_cust.dtypes

user_id                  int64
First Name              object
Surnam                  object
Gender                  object
STATE                   object
Age                      int64
date_joined             object
number_of_dependants     int64
family_status           object
income                   int64
dtype: object

In [16]:
# Change date_joined column data type from object to datetime64
df_cust['date_joined'] = df_cust['date_joined'].astype('datetime64')

In [17]:
df_cust.dtypes

user_id                          int64
First Name                      object
Surnam                          object
Gender                          object
STATE                           object
Age                              int64
date_joined             datetime64[ns]
number_of_dependants             int64
family_status                   object
income                           int64
dtype: object

### Changed date_joined column data type from object to datetime64

In [23]:
# Check values
df_cust['Gender'].value_counts(dropna = False)

Male      104067
Female    102142
Name: Gender, dtype: int64

In [24]:
# Check values
df_cust['STATE'].value_counts(dropna = False)

Florida                 4044
Colorado                4044
Illinois                4044
Alabama                 4044
District of Columbia    4044
Hawaii                  4044
Arizona                 4044
Connecticut             4044
California              4044
Indiana                 4044
Arkansas                4044
Alaska                  4044
Delaware                4044
Iowa                    4044
Idaho                   4044
Georgia                 4044
Wyoming                 4043
Mississippi             4043
Oklahoma                4043
Utah                    4043
New Hampshire           4043
Kentucky                4043
Maryland                4043
Rhode Island            4043
Massachusetts           4043
Michigan                4043
New Jersey              4043
Kansas                  4043
South Dakota            4043
Minnesota               4043
Tennessee               4043
New York                4043
Washington              4043
Louisiana               4043
Montana       

In [31]:
# Check values
df_cust['Age'].min()

18

In [32]:
# Check values
df_cust['Age'].max()

81

In [35]:
# Check values
df_cust['date_joined'].max()

Timestamp('2020-04-01 00:00:00')

In [36]:
# Check values
df_cust['date_joined'].min()

Timestamp('2017-01-01 00:00:00')

In [37]:
# Check values
df_cust['number_of_dependants'].value_counts(dropna = False)

0    51602
3    51594
1    51531
2    51482
Name: number_of_dependants, dtype: int64

In [38]:
# Check values
df_cust['family_status'].value_counts(dropna = False)

married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: family_status, dtype: int64

In [40]:
# Check values
df_cust['income'].max()

593901

In [41]:
# Check values
df_cust['income'].min()

25903

In [42]:
# Check for duplicate rows
df_cust_dups = df_cust[df_cust.duplicated()]

In [43]:
df_cust_dups

Empty DataFrame
Columns: [user_id, First Name, Surnam, Gender, STATE, Age, date_joined, number_of_dependants, family_status, income]
Index: []

### No duplicate rows returned

In [46]:
# Export cleaned customer dataset as pickle
df_cust.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'customers_checked.pkl'))

In [55]:
df_cust.shape

(206209, 10)

# SAVE AND CHECKPOINT

## 6. Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)


In [48]:
# Merge df_cust and ords_prods_merge
cust_ords_prods_merge = df_cust.merge(ords_prods_merge, on = 'user_id')

In [49]:
cust_ords_prods_merge

user_id First Name    Surnam  Gender     STATE  Age date_joined  \
0           26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
1           26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
2           26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
3           26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
4           26711    Deborah  Esquivel  Female  Missouri   48  2017-01-01   
...           ...        ...       ...     ...       ...  ...         ...   
32404854    80148    Cynthia     Noble  Female  New York   55  2020-04-01   
32404855    80148    Cynthia     Noble  Female  New York   55  2020-04-01   
32404856    80148    Cynthia     Noble  Female  New York   55  2020-04-01   
32404857    80148    Cynthia     Noble  Female  New York   55  2020-04-01   
32404858    80148    Cynthia     Noble  Female  New York   55  2020-04-01   

          number_of_dependants family_status  income  ...  prices  \
0                            3       married  165665  ...     9.0   
1                            3       married  165665  ...     9.0   
2                            3       married  165665  ...     9.0   
3                            3       married  165665  ...    12.6   
4                            3       married  165665  ...    12.6   
...                        ...           ...     ...  ...     ...   
32404854                     1       married   57095  ...     4.7   
32404855                     1       married   57095  ...     4.0   
32404856                     1       married   57095  ...     4.7   
32404857                     1       married   57095  ...     1.8   
32404858                     1       married   57095  ...     6.0   

            price_range_loc        busiest_days  busiest_period_of_day  \
0         Mid-range product        Busiest days            Most orders   
1         Mid-range product      Regularly busy            Most orders   
2         Mid-range product        Busiest days            Most orders   
3         Mid-range product      Regularly busy            Most orders   
4         Mid-range product  Least busiest days            Most orders   
...                     ...                 ...                    ...   
32404854  Low-range product      Regularly busy            Most orders   
32404855  Low-range product      Regularly busy            Most orders   
32404856  Low-range product      Regularly busy         Average orders   
32404857  Low-range product      Regularly busy            Most orders   
32404858  Mid-range product      Regularly busy            Most orders   

          max_order  loyalty_flag  avg_price_per_order  spending_flag  \
0                 8  New customer             7.988889    Low spender   
1                 8  New customer             7.988889    Low spender   
2                 8  New customer             7.988889    Low spender   
3                 8  New customer             7.988889    Low spender   
4                 8  New customer             7.988889    Low spender   
...             ...           ...                  ...            ...   
32404854          4  New customer             3.886667    Low spender   
32404855          4  New customer             3.886667    Low spender   
32404856          4  New customer             3.886667    Low spender   
32404857          4  New customer             3.886667    Low spender   
32404858          4  New customer             3.886667    Low spender   

         median_days_since_prior_order order_frequency_flag  
0                                 19.0     Regular customer  
1                                 19.0     Regular customer  
2                                 19.0     Regular customer  
3                                 19.0     Regular customer  
4                                 19.0     Regular customer  
...                                ...                  ...  
32404854                          12.0     Regular custo

In [50]:
# Check merged value counts
cust_ords_prods_merge['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [51]:
cust_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 32 columns):
 #   Column                         Dtype         
---  ------                         -----         
 0   user_id                        int64         
 1   First Name                     object        
 2   Surnam                         object        
 3   Gender                         object        
 4   STATE                          object        
 5   Age                            int64         
 6   date_joined                    datetime64[ns]
 7   number_of_dependants           int64         
 8   family_status                  object        
 9   income                         int64         
 10  order_id                       int64         
 11  user_order_number              int64         
 12  orders_day_of_week             int64         
 13  order_hour_of_day              int64         
 14  days_since_prior_order         float64       
 15  product_id   

In [52]:
cust_ords_prods_merge.shape

(32404859, 32)

# SAVE AND CHECKPOINT

## 8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [53]:
# Export cust_ords_prods_merge dataset as pickle
cust_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'cust_orders_products_merged_Task_4_9.pkl'))